dynamic_array/include/dynamic_array/array.h

```diff
@@ -0,0 +1,303 @@
+}
+
+template <typename T>
+Array<T>& Array<T>::operator=(const Array<T>& other) {
+  Array<T> temp_array(other);
+  swap(*this, temp_array);
+  return *this;
+}
+
+template <typename T>
+Array<T>& Array<T>::operator=(Array<T>&& other) {
+  // Maybe shouldn't use |swap| method for move assignment operator
+  // https://stackoverflow.com/questions/6687388/why-do-some-people-use-swap-for-move-assignments
+  Array<T> temp_array = std::move(other);
+  swap(*this, temp_array);
+  return *this;
+}
+
+template <typename T>
+Array<T>::~Array() {
+  delete[] items_;
+}
+
+template <typename T>
+size_t Array<T>::size() {
+  return size_;
+}
+
+template <typename T>
+size_t Array<T>::capacity() {
+  return capacity_;
+}
+
+template <typename T>
+bool Array<T>::is_empty() {
+  return size_ == 0;
+}
+
+template <typename T>
+T Array<T>::item_at(int index) {
+  throw_out_of_range_error_if_needed(index);
+
+  return items_[index];
+}
+
+template <typename T>
+void Array<T>::append(const T& item) {
+  reallocate_if_needed(size_ + 1);
+
+  items_[size_] = item;
+  size_++;
+}
+
+template <typename T>
+void Array<T>::insert(const T& item, int index) {
+  throw_out_of_range_error_if_needed(index);
+  reallocate_if_needed(size_ + 1);
+
+  for (size_t i = size_; i > index; i--) {
+    items_[i] = items_[i - 1];
+  }
+
+  items_[index] = item;
+  size_++;
+}
+
+template <typename T>
+void Array<T>::prepend(const T& item) {
+  insert(item, 0);
+}
+
+template <typename T>
+T Array<T>::pop() {
+  throw_out_of_range_error_if_needed(0);
+
+  T& last_item = items_[size_ - 1];
+
+  reallocate_if_needed(--size_);
+
+  return last_item;
+}
+
+template <typename T>
+void Array<T>::remove_at(int index) {
+  throw_out_of_range_error_if_needed(index);
+
+  for (size_t i = index; i < size_ - 1; i++) {
+    items_[i] = items_[i + 1];
+  }
+
+  reallocate_if_needed(--size_);
+}
+
+template <typename T>
+void Array<T>::remove(const T& item) {
+  for (size_t i = 0; i < size_; i++) {
+    if (items_[i] != item) {
+      continue;
+    }
+
+    remove_at(i);
+    i--;
+  }
+}
+
+template <typename T>
+int Array<T>::find(const T& item) {
+  for (size_t i = 0; i < size_; i++) {
+    if (item == items_[i]) {
+      return i;
+    }
+  }
+
+  return index_not_found;
+}
+
+// Private
+template <typename T>
+void Array<T>::throw_out_of_range_error_if_needed(int index) {
+  if (!is_empty() && index >= 0 && index < size_) {
+    return;
+  }
+
+  throw std::out_of_range("Index " + std::to_string(index) + " out of bounds " +
+                          std::to_string(size_) + ".");
+}
+
+template <typename T>
+void Array<T>::reallocate_if_needed(size_t new_size) {
+  size_t new_capacity = capacity_;
+
+  if (new_size > capacity_) {
+    new_capacity *= growth_factor;
+  } else if (new_size <= capacity_ / shrink_factor &&
+             capacity_ / growth_factor >= min_capacity) {
+    new_capacity /= growth_factor;
+  } else {
+    return;
+  }
+
+  T* new_items = new T[new_capacity];
```

Comment:
Where does it belong to?

It is not different with `T new_items[new_capacity]`